azkaban-common/src/main/java/azkaban/metric/MetricReportManager.java

```diff
@@ -0,0 +1,220 @@
+  private static boolean _isManagerEnabled;
+
+  private MetricReportManager() {
+    _logger.debug("Instantiating Metric Manager");
+    _executorService = Executors.newFixedThreadPool(MAX_EMITTER_THREADS);
+    _metrics = new ArrayList<IMetric<?>>();
+    _metricEmitters = new LinkedList<IMetricEmitter>();
+    enableManager();
+  }
+
+  /**
+   * @return true, if we have enabled metric manager from Azkaban exec server
+   */
+  public static boolean isAvailable() {
+    return _instance != null && _isManagerEnabled;
+  }
+
+  /**
+   * Get a singleton object for Metric Manager
+   */
+  public static MetricReportManager getInstance() {
+    if (_instance == null) {
+      synchronized (MetricReportManager.class) {
+        if (_instance == null) {
+          _logger.info("Instantiating MetricReportManager");
+          _instance = new MetricReportManager();
+        }
+      }
+    }
+    return _instance;
+  }
+
+  /***
+   * each element of metrics List is responsible to call this method and report metrics
+   * @param metric
+   */
+  public void reportMetric(final IMetric<?> metric) {
+    if (metric != null && isAvailable()) {
+      try {
+        final IMetric<?> metricSnapshot;
+        // take snapshot
+        synchronized (metric) {
+          metricSnapshot = metric.getSnapshot();
+        }
+        _logger.debug(String.format("Submitting %s metric for metric emission pool", metricSnapshot.getName()));
+        // report to all emitters
+        for (final IMetricEmitter metricEmitter : _metricEmitters) {
+          _executorService.submit(new Runnable() {
+            @Override
+            public void run() {
+              try {
+                metricEmitter.reportMetric(metricSnapshot);
+              } catch (Exception ex) {
+                _logger.error(String.format("Failed to report %s metric due to %s", metricSnapshot.getName(),
+                    ex.toString()));
+              }
+            }
+          });
+        }
+      } catch (CloneNotSupportedException ex) {
+        _logger.error(String.format("Failed to take snapshot for %s metric %s", metric.getClass().getName(),
+            ex.toString()));
+      }
+    }
+  }
+
+  /**
+   * Add a metric emitter to report metric
+   * @param emitter
+   */
+  public void addMetricEmitter(final IMetricEmitter emitter) {
+    _metricEmitters.add(emitter);
+  }
+
+  /**
+   * remove a metric emitter
+   * @param emitter
+   */
+  public void removeMetricEmitter(final IMetricEmitter emitter) {
+    _metricEmitters.remove(emitter);
+  }
+
+  /**
+   * Get all the metric emitters
+   * @return
+   */
+  public List<IMetricEmitter> getMetricEmitters() {
+    return _metricEmitters;
+  }
+
+  /**
+   * Add a metric to be managed by Metric Manager
+   * @param metric
+   */
+  public void addMetric(final IMetric<?> metric) {
+    // metric null or already present
+    if (metric != null && getMetricFromName(metric.getName()) == null) {
```

Comment:
Does it make sense to separate this out into two cases?
1) Argument metric is null -> throw IllegalArgumentException
2) getMetricFromName(metric.getName()) == null